ceno_zkvm/src/scheme/verifier.rs

```diff
@@ -140,43 +140,63 @@ impl<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>> ZKVMVerifier<E, PCS>
         let point_eval = PointAndEval::default();
         let mut transcripts = transcript.fork(self.vk.circuit_vks.len());
 
-        for (name, (i, opcode_proof)) in vm_proof.opcode_proofs {
-            let transcript = &mut transcripts[i];
+        // For each opcode, include the num_instances 
+        // into its corresponding fork of the transcript.
+        for ((name, _), (i, transcript)) in self
+            .vk
+            .circuit_vks
+            .iter() // Sorted by key.
+            .zip_eq(transcripts.iter_mut().enumerate())
```

Comment:
It's probably less confusing to do the enumerate once on the outside of the zip:

```diff
diff --git a/ceno_zkvm/src/scheme/verifier.rs b/ceno_zkvm/src/scheme/verifier.rs
index 52c11a2c..d6b3e41a 100644
--- a/ceno_zkvm/src/scheme/verifier.rs
+++ b/ceno_zkvm/src/scheme/verifier.rs
@@ -4,7 +4,7 @@ use ark_std::iterable::Iterable;
 use ceno_emul::WORD_SIZE;
 use ff_ext::ExtensionField;
 
-use itertools::{Itertools, interleave, izip};
+use itertools::{Itertools, enumerate, interleave, izip};
 use mpcs::PolynomialCommitmentScheme;
 use multilinear_extensions::{
     mle::{IntoMLE, MultilinearExtension},
@@ -142,12 +142,8 @@ impl<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>> ZKVMVerifier<E, PCS>
 
         // For each opcode, include the num_instances
         // into its corresponding fork of the transcript.
-        for ((name, _), (i, transcript)) in self
-            .vk
-            .circuit_vks
-            .iter() // Sorted by key.
-            .zip_eq(transcripts.iter_mut().enumerate())
-        {
+        // Note: circuits are sorted by key.
+        for (i, ((name, _), transcript)) in enumerate(izip!(self.vk.circuit_vks, transcripts)) {
             // get num_instances from opcode proof
             let opcode_result = vm_proof.opcode_proofs.get(name);
             let num_instances = opcode_result.map(|(_, p)| p.num_instances).unwrap_or(0);
```

Reply:
Enumeration is not needed just for updating the transcripts, so I removed it. 

Just a remark: without "iter_mut()" the compiler asks to clone the transcripts, which is not what we want, right?  So maybe I am missing something from your suggestion here. Any way, since the enumeration was removed, this discussion may be useless now. But I appreciate your comment, thank you.